fluent-plugin-enhance-k8s-metadata/lib/sumologic/kubernetes/service_monitor.rb

```diff
@@ -0,0 +1,132 @@
+        log.debug "Starting service endpoints watcher thread"
+        params = Hash.new
+        params[:as] = :raw
+        params[:resource_version] = get_current_service_snapshot_resource_version
+        params[:timeout_seconds] = @watch_service_interval_seconds + 60
+
+        @watcher = @clients['v1'].public_send("watch_endpoints", params).tap do |watcher|
+          thread_create(:"watch_endpoints") do
+            @watch_stream = watcher
+            @watcher_id = Thread.current.object_id
+            log.debug "New thread to watch service endpoints #{@watcher_id} from resource version #{params[:resource_version]}"
+
+            watcher.each do |event|
+              begin
+                event = JSON.parse(event)
+                handle_service_event(event)
+              rescue => e
+                log.error "Got exception #{e} parsing entity #{entity}. Skipping."
+              end
+            end
+            log.info "Closing watch stream"
+          end
+        end
+      end
+
+      def get_current_service_snapshot_resource_version
+        log.debug "Getting current service snapshot"
+        begin
+          params = Hash.new
+          params[:as] = :raw
+          response = @clients['v1'].public_send "get_endpoints", params
+          result = JSON.parse(response)
+          new_snapshot_pods_to_services = Concurrent::Map.new {|h, k| h[k] = []}
+
+          result['items'].each do |endpoint|
+            service = endpoint['metadata']['name']
+            get_pods_for_service(endpoint).each {|pod| new_snapshot_pods_to_services[pod] << service}
+          end
+
+          @pods_to_services = new_snapshot_pods_to_services
+          result['metadata']['resourceVersion']
+        rescue => e
+          log.error "Got exception #{e} getting current service snapshot and corresponding resource version."
+          0
+        end
+      end
+
+      def get_pods_for_service(endpoint)
+        pods = []
+        if endpoint.key? 'subsets'
+          endpoint['subsets'].each do |subset|
+            ['addresses', 'notReadyAddresses'].each do |key|
+              if subset.key? key
+                subset[key].each do |object|
+                  if object.key? 'targetRef'
+                    if object['targetRef']['kind'] == 'Pod'
+                      pod = object['targetRef']['name']
+                      log.debug "Found Pod #{pod} for Service #{endpoint['metadata']['name']}"
+                      pods << pod
+                    end
+                  end
+                end
+              end
+            end
+          end
+        end
+        pods
+      end
+
+      def handle_service_event(event)
+        type = event['type']
+        endpoint = event['object']
+        service = endpoint['metadata']['name']
+        case type
+        when 'ADDED'
+          get_pods_for_service(endpoint).each {|pod| @pods_to_services[pod] << service unless @pods_to_services[pod].include? service}
+        when 'MODIFIED'
```

Comment:
not sure if I'm following the `MODIFIED` case, can you give some explanation?

Reply:
Yeah, the `MODIFIED` case was the hardest part. There's two cases to cover:
1. The service was modified such that there are more pods than there once were (increasing the number of replicas). This is the easier case. We simply add the service to `@pods_to_services[pod]` unless it already exists.
2. The service was modified such that there are now fewer pods than there once were (downscaling the number of replicas). This is the harder case, since you don't know which pods were removed you have to actually search the hash for mentions of the service, and delete it if it's not one of the pods we were told about in the `MODIFIED` event. Then if there are no services for that pod in the hash, this means the pod is deleted itself so we can remove that key from the hash.

I offhandedly asked Chris about it at some point, since the "looking through the entire hash for mentions of the service" part really irked me... but he seemed to think it wasn't that big of an issue to be worth "over-optimizing prematurely". I'm open to any suggestions 😅

Reply:
thanks for the detailed explanation! I'm fine with current approach